crates/sui-verifier/src/char_type_verifier.rs

```diff
@@ -0,0 +1,231 @@
+            // characteristic type candidate
+            verify_no_instantiations(module, fn_def, candidate_name, def)
+                .map_err(verification_failure)?;
+        }
+    }
+
+    Ok(())
+}
+
+// Verifies all required properties of a characteristic type candidate (that is a type whose name is
+// the same as the name of a
+fn verify_char_type(
+    module: &CompiledModule,
+    struct_name: &str,
+    struct_handle: &StructHandle,
+    struct_def: &StructDefinition,
+) -> Result<(), String> {
+    // must have only one ability: drop
+    let drop_set = AbilitySet::EMPTY | Ability::Drop;
+    let abilities = struct_handle.abilities;
+    if abilities != drop_set {
+        return Err(format!(
+            "characteristic type candidate {}::{} must have a single ability: drop",
+            module.self_id(),
+            struct_name,
+        ));
+    }
+    let field_count = struct_def.declared_field_count().map_err(|_| {
+        format!(
+            "characteristic type candidate {}::{} cannot be a native structure",
+            module.self_id(),
+            struct_name
+        )
+    })?;
+
+    // unwrap below is safe as it will always be successful if declared_field_count call above is
+    // successful
+    if field_count != 1 || struct_def.field(0).unwrap().signature.0 != SignatureToken::Bool {
+        return Err(format!(
+            "characteristic type candidate {}::{} must have a single bool field only (or no fields)",
+            module.self_id(),
+            struct_name,
+        ));
+    }
+
+    if !struct_handle.type_parameters.is_empty() {
+        return Err(format!(
+            "characteristic type candidate {}::{} cannot have type parameters",
+            module.self_id(),
+            struct_name,
+        ));
+    }
+    Ok(())
+}
+
+/// Checks if this module's `init` function conformant with the characteristic type
+fn verify_init_function_char_type(
+    module: &CompiledModule,
+    fn_handle: &FunctionHandle,
+    struct_name: &str,
+    struct_handle: &StructHandle,
+) -> Result<(), String> {
+    let view = &BinaryIndexedView::Module(module);
+    let fn_sig = view.signature_at(fn_handle.parameters);
+    if fn_sig.len() != 2 || !is_char_type(view, &fn_sig.0[0], struct_handle) {
+        // check only the first parameter - the other one is checked in entry_points verification
+        // pass
+        return Err(format!(
+            "init function of a module containing characteristic type candidate must have {}::{} as the first parameter",
+            module.self_id(),
+            struct_name,
+        ));
+    }
+
+    Ok(())
+}
+
+// Checks if a given SignatureToken represents a characteristic type struct
+fn is_char_type(
+    view: &BinaryIndexedView,
+    tok: &SignatureToken,
+    struct_handle: &StructHandle,
+) -> bool {
+    if let SignatureToken::Struct(idx) = tok {
+        if view.struct_handle_at(*idx) == struct_handle {
```

Comment:
if you want to be fancy, `matches(tok, SignatureToken::Struct(idx) if view.struct_handle_at(*idx) == struct_handle)

Reply:
It is fancy indeed!